Bootstrap 5 upgrade without theme feature.

gammamatrix opened the pull request:
## Changes
- The theme feature was removed in this branch.
- See GH-1036

sebastianbergmann commented:
Are these visual changes intentional?

#### Before these changes (cafa435d750d52e5e7c9c671341483d12d260c45)
![grafik](https://github.com/sebastianbergmann/php-code-coverage/assets/25218/560e105a-568d-4549-a189-07a279350b2b)

#### After these changes (ba3047f872113a15df70f8e5d246b0524681f98c)
![grafik](https://github.com/sebastianbergmann/php-code-coverage/assets/25218/2c9c72f4-be0d-4a55-8755-8b2b40bacad5)

gammamatrix commented:
> Are these visual changes intentional?
> 
No. I am looking into this now. It could be browser specific.

I am reviewing the rules in style.css; the old rules definitely need a bit of modification due to a new Bootstrap 5 rule for tables:
```css
.table>:not(caption)>*>* {
    padding: .5rem .5rem;
    color: var(--bs-table-color-state,var(--bs-table-color-type,var(--bs-table-color)));
    background-color: var(--bs-table-bg);
    border-bottom-width: var(--bs-border-width);
    box-shadow: inset 0 0 0 9999px var(--bs-table-bg-state,var(--bs-table-bg-type,var(--bs-table-accent-bg)))
}
```
- This rule changes the `td background-color`.
- I will test a few more browsers and operating systems to make sure the colors propagate from the `table.tr` to `table.tr.td`, as expected.
- I have at least one minor change so far. I think I may have to alter a few more rules, once I test a bit.

gammamatrix commented:
## Expected output

I updated a few rules, and undid a few modified rules.
- I have not tested other browsers yet, it is possible there is an issue with the CSS function for `rgb(from ...)`:
```css
rgb(from var(--bs-warning) r g b / 0.1)
```
This is what I am seeing with the latest changes:
![php-code-coverage-bootstrap-5-GH-1037](https://github.com/sebastianbergmann/php-code-coverage/assets/524724/cae66eb8-f36c-48ac-99a4-7af21f6a04a4)

gammamatrix commented:
### Findings

It does appear to be a browser compatibility issue with only Firefox for relative colors: https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/rgb#browser_compatibility

I went ahead and put back the original default colors into `Colors.php`:
```php
return new self('#dff0d8', '#c3e3b5', '#99cb84', '#fcf8e3', '#f2dede');
```
The theme friendly rules can still be applied as environment variables:
```xml
<html outputDirectory="output/html" lowUpperBound="50" highLowerBound="90"
    colorSuccessLow="rgb(from var(--bs-purple) r g b / 0.75)"
    colorSuccessMedium="rgb(from var(--bs-blue) r g b / 0.55)"
    colorSuccessHigh="rgb(from var(--bs-teal) r g b / 0.25)"
    colorWarning="rgb(from var(--bs-warning) r g b / 0.25)"
    colorDanger="rgb(from var(--bs-danger) r g b / 0.25)" />
```

**NOTE:** Mozilla has approved relative colors, but it has not made it into Firefox yet: 
- https://github.com/mozilla/standards-positions/issues/841

sebastianbergmann commented:
Thank you for working on this. Unfortunately, I found more visual differences:

Are these visual changes intentional?

#### Before these changes (cafa435d750d52e5e7c9c671341483d12d260c45)
![grafik](https://github.com/sebastianbergmann/php-code-coverage/assets/25218/5f9e95c5-5a60-46dd-8b9d-f6e642a87eff)

#### After these changes (3ab35d131e572e50856f37d055db10de34673c4a)
![grafik](https://github.com/sebastianbergmann/php-code-coverage/assets/25218/d5a716e6-9d04-4bd3-b2d1-5b2e4e8a317a)

gammamatrix commented:
> Thank you for working on this. Unfortunately, I found more visual differences:
> 
> Are these visual changes intentional?
> 
No problem, happy to help! And no, those changes are not intentional.

In this case, the Bootstrap 5 rules are taking precedence with the use of: `bg-success, bg-warning, bg-danger`

```css
.bg-success {
  --bs-bg-opacity: 1;
  background-color: rgba(var(--bs-success-rgb),var(--bs-bg-opacity)) !important;
}
```
- I will take a look and see if there are some simple CSS modifications I can make to style.css.

![php-code-coverage-bootstrap-progress-bars](https://github.com/sebastianbergmann/php-code-coverage/assets/524724/740af002-8b92-4e92-98ad-f2954be7faba)

gammamatrix commented:
## Findings for progress-bar background colors

So it looks like the optional colors have never overridden the progress-bar. Bootstrap 4 just used a lighter color for `bg-success`:
```css
.bg-success {
    background-color: #28a745 !important;
}
```

![php-code-coverage-bootstrap-4-used-different-green](https://github.com/sebastianbergmann/php-code-coverage/assets/524724/096dc2a6-8069-435f-8b2c-824ffa904e14)

I was experimenting with some rules to allow the progress bars to override colors.

This is what it would look like with Bootstrap 4 and the override colors, which is not ideal:

![php-code-coverage-bootstrap-4-override-progress-bar-colors](https://github.com/sebastianbergmann/php-code-coverage/assets/524724/33fe3e93-b25b-4a73-bfc2-1a39d5025406)

I have a few more things I can try, but a final solution might require the usage of themes along with `CSS rgba()`. If it is too complicated or messy, I will update this comment with details so we can figure out how to proceed.

gammamatrix commented:
## Proposed final changes (without data-bs-theme variable)

**FYI:** *Still testing - I think I have a working solution - will update this comment with more details*

In this last pass, I added root variables for `--phpunit` at the top of the style.css file:

```css
/* :root {
 --phpunit-success-bar: #28a745;
 --phpunit-success-high: #99cb84;
 --phpunit-success-medium: #c3e3b5;
 --phpunit-success-low: #dff0d8;
 --phpunit-warning: #fcf8e3;
 --phpunit-warning-bar: #ffc107;
 --phpunit-danger: #f2dede;
 --phpunit-danger-bar: #dc3545;
} */

:root {
 --phpunit-success-bar: #28a745;
 --phpunit-success-high: {{success-high}};
 --phpunit-success-medium: {{success-medium}};
 --phpunit-success-low: {{success-low}};
 --phpunit-warning: {{warning}};
 --phpunit-warning-bar: #ffc107;
 --phpunit-danger: {{danger}};
 --phpunit-danger-bar: #dc3545;
}
```
- the existing color options get set as before, with no changes to `Colors.php`. I added a few `--phpunit-*-bar` rules to cover the old Bootstrap 4 color so it looks the same.
- So far, I have tested overriding the defined `--phpunit-*` in the `customCssFile` and that seems to be working properly in Firefox as well for the dark mode theme.
- **NOTE:** There is a unit test that fails saying the files do not match:
```
There was 1 failure:

SebastianBergmann\CodeCoverage\Report\Html\EndToEndTest::testPathCoverageForBankAccountTest
BankAccount.php_path.html not match
Failed asserting that string matches format description.
--- Expected
+++ Actual

/Users/jeremy/srv/sites/site-playground-make/vendor/gammamatrix/playground-make-controller/vendor/phpunit/php-code-coverage/tests/tests/Report/Html/EndToEndTest.php:108
/Users/jeremy/srv/sites/site-playground-make/vendor/gammamatrix/playground-make-controller/vendor/phpunit/php-code-coverage/tests/tests/Report/Html/EndToEndTest.php:56
```

sebastianbergmann commented:
I just tested with the current state of the `update-to-bootstrap-5-without-theme-feature` branch (at ea94640a1a7dd6a8cd9eaa5e752a74e3af57ef71) and I still see the large amount of whitespace on each line between the line number and the code that I posted a screenshot of in https://github.com/sebastianbergmann/php-code-coverage/pull/1037#issuecomment-2106675841.

Sorry for not getting back to this sooner.

gammamatrix commented:
> I just tested with the current state of the `update-to-bootstrap-5-without-theme-feature` branch (at [ea94640](https://github.com/sebastianbergmann/php-code-coverage/commit/ea94640a1a7dd6a8cd9eaa5e752a74e3af57ef71)) and I still see the large amount of whitespace on each line between the line number and the code that I posted a screenshot of in [#1037 (comment)](https://github.com/sebastianbergmann/php-code-coverage/pull/1037#issuecomment-2106675841).
> 
> Sorry for not getting back to this sooner.

No worries, I will update my local branches and have a look at the tests to see why the extra whitespace is appearing in the test.

sebastianbergmann commented:
My bad: I just noticed that the large amount of whitespace between the left "edge" and the line number also exists in the current releases and therefore cannot be caused by this PR.

sebastianbergmann commented:
Here is what I currently see:

### `main` (at e73593b1b88df891ed17cee70b13b97bdcadc275)
#### Directory
![grafik](https://github.com/user-attachments/assets/52bb2673-d64f-44b6-95e9-57a181181089)
#### File
![grafik](https://github.com/user-attachments/assets/76755a9c-efb8-49aa-870f-a4ea5e0342fc)
### `update-to-bootstrap-5-without-theme-feature` (at ea94640a1a7dd6a8cd9eaa5e752a74e3af57ef71)
#### Directory
![grafik](https://github.com/user-attachments/assets/c1443796-3395-4c68-b206-8dfb0a78b554)
#### File
![grafik](https://github.com/user-attachments/assets/bd9aecae-6e63-4ff5-8d5a-f0febba11f7d)

Can we restore the styling of the top navigation / breadcrumbs and reduce the whitespace between the line numbers and the beginning of the lines? Thanks!

gammamatrix commented:
## Breadcrumbs Changes
- I changed the background color (added by Bootstrap) to transparency. I think this should be theme and customization safe, but did not test yet.

These style changes are visible in the inspector on the screenshot:
```css
nav .breadcrumb {
  background-color: transparent !important;
  margin: 0;
  padding-left: 0;
  padding-top: 0;
}
```
![php-code-coverage-GH-1037-breadcrumbs](https://github.com/user-attachments/assets/982e84bd-8ec5-475f-ac64-e202fecd1938)
- From Chrome, looked similar on Firefox.

**Status** Not committed yet. Going to tackle the line numbers next.

gammamatrix commented:
> Sorry for bothering you with this, but after merging this I see the whitespace left of the line numbers again:
> 
> ![grafik](https://private-user-images.githubusercontent.com/25218/364831174-65a71cfd-0803-48b5-8927-28d6ba8feb39.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU1NTk4ODYsIm5iZiI6MTcyNTU1OTU4NiwicGF0aCI6Ii8yNTIxOC8zNjQ4MzExNzQtNjVhNzFjZmQtMDgwMy00OGI1LTg5MjctMjhkNmJhOGZlYjM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA1VDE4MDYyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiOTgxZTExYTQ4MjMyMjYwZTk2OGY0NGQwNWI3NjE1MzM4ZTFkOTY5ZjY5YjIzOGM4ZDk0MWYxNmU3MDI5NGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V9NtszN9SklSubXQ2Glq8uD11DOCyRElNLgV63aNtDo)
> 
> Also, I think the line numbers were not underlined in the past.

I will have a look.

gammamatrix commented:
## Dev Notes
I was able to get a rule that works:
```css
table#code {
  td:nth-child(1) {
    padding-left: .75em;
    padding-right: .75em;
    a {
      text-decoration: none;
     }
  }
}
```
- Bootstrap 5 no longer adds `text-decoration: none;` to links.

![GH-1037-css-fix-for-first-column](https://github.com/user-attachments/assets/ee4f2e59-c928-4c80-9c38-121e80a2ed5a)

I will open up a new PR instead of using this branch.

gammamatrix commented:
## Changes
- See: https://github.com/sebastianbergmann/php-code-coverage/pull/1046
